### PR TITLE
Additional validation on updating social fields via API v2 [PLAT-925]

### PR DESCRIPTION
**Pull request by erinspace**
## Purpose

One could update social fields via API v2 after the social fields were added. This adds additional validation using jsonschema, and ~~allows fields that aren't currently allowed for in the model to be sent and ignored.~~

This updates the format a bit, migrates old data to match that format, and makes sure that payloads to set the data conform to that spec.

## Changes

- add jsonschema document for the social fields
- use this to validate data being sent to the API
- add a validate method on the serializer to process the data
- Social fields that can be lists are now stored as lists, and those that are singular are stored as strings

## QA Notes

This would only require API testing, and shouldn't be very different from what exists already anyhow! Should be all good, and will be tested by Embosf when they use this feature for the social page!

## Documentation
https://github.com/CenterForOpenScience/developer.osf.io/pull/23

## Side Effects
None anticipated

## Ticket

https://openscience.atlassian.net/browse/PLAT-925

### CODE REVIEW

File: api/users/schemas.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+
+def get_user_social_jsonschema():
```
Comment (pattisdr) [inline api/users/schemas.py]: Nice schema 👍

File: api/users/serializers.py [hunk 2]
```diff
@@ -128,20 +130,31 @@ def profile_image_url(self, user):
         size = self.context['request'].query_params.get('profile_image_size')
         return user.profile_image_url(size=size)
 
+    def validate_social(self, value):
```
Comment (pattisdr) [inline api/users/serializers.py]: I forgot you could do this! `.validate_<field_name> `

File: api/users/schemas.py (hunk 1 above)
Comment (caseyrollins) [inline api/users/schemas.py]: @erinspace, @Johnetordoff's PR adds validation schemas for employment and education information and [implements a `from_json` method](https://github.com/CenterForOpenScience/osf.io/pull/8546/files#diff-517a736b64468905bd7818494c33efdaR1) for loading the correct schema. Would you mind mimicking that approach in this PR?

File: api/users/serializers.py [hunk 3]
```diff
@@ -128,20 +130,28 @@ def profile_image_url(self, user):
...
+        schema = from_json('social-schema.json')
+        try:
+            jsonschema.validate(value, schema)
+        except jsonschema.ValidationError as e:
+            raise InvalidModelValueError(e)
+
+        social_dict = {}
+        for key, val in value.iteritems():
+            # Ignore fields that are not specified in the current social fields list
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/users/serializers.py]: Just for history's sake, when I wrote this the requirement was "Ignore attempts to update social fields that are not in [the list of fields]" which is what this was doing. Requirement has changed so will now error with extra fields instead of ignoring.

**Comment by sloria** [comment]
@erinspace Can you get this up to date with develop and repoint the PR pls?

**Comment by erinspace** [comment]
@sloria, up-to-dated and re-pointed!

File: osf/migrations/0112_update_social_data_format.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    users_to_update = users_with_social.count()
+    logger.info('Updating social fields for {} users'.format(users_to_update))
+    progress_bar = progressbar.ProgressBar(maxval=users_to_update or 100).start()
+
+    users_updated = 0
+    for user in users_with_social:
+        old_social = {}
+        for key, value in user.social.iteritems():
+            if key in SOCIAL_LIST_FIELDS and key != 'profileWebsites':
+                    old_social[key] = value[0]
```
Comment (sloria) [inline osf/migrations/0112_update_social_data_format.py]: The migration should error if it finds a value that contains more than one item, so we don't accidentally remove data.

File: api/users/schemas.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import os
+import json
+
+here = os.path.split(os.path.abspath(__file__))[0]
+
+def from_json(fname):
```
Comment (sloria) [inline api/users/schemas.py]: This file was moved to a directory. I'm getting an `ImportError` when initializing the app. I think this file just needs to be removed.

File: api/users/serializers.py [hunk 6]
```diff
@@ -1,19 +1,21 @@
...
     VersionedDateTimeField, HideIfDisabled, IDField,
-    Link, LinksField, ListDictField, TypeField, RelationshipField, JSONAPIListField,
+    Link, LinksField, TypeField, RelationshipField, JSONAPIListField,
     WaterbutlerLink, ShowIfCurrentUser
 )
 from api.base.utils import absolute_reverse, get_user_auth, waterbutler_api_url_for
 from api.files.serializers import QuickFilesSerializer
 from osf.exceptions import ValidationValueError, ValidationError
 from osf.models import OSFUser, QuickFilesNode
+from api.users.schemas import from_json
```
Comment (sloria) [inline api/users/serializers.py]: should be `api.users.schemas.utils`.

File: osf/migrations/0112_update_social_data_format.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+def reset_social_fields(state, schema):
+    OSFUser = state.get_model('osf', 'osfuser')
+    users_with_social = OSFUser.objects.exclude(social={})
+    users_to_update = users_with_social.count()
+    logger.info('Updating social fields for {} users'.format(users_to_update))
+    progress_bar = progressbar.ProgressBar(maxval=users_to_update or 100).start()
+
+    users_updated = 0
+    for user in users_with_social:
+        old_social = {}
```
[the remark this answers is not in the document]
Comment (erinspace) [reply osf/migrations/0112_update_social_data_format.py]: Naw, cause it's in the `reset_social_fields` method, so it's putting it back to the old social field!

File: osf/migrations/0124_update_social_data_format.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+
+FIELDS_TO_MIGRATE = [
+    'github',
+    'linkedIn',
+    'twitter'
+]
+
+
+BATCH_SIZE = 1000
```
Comment (sloria) [inline osf/migrations/0124_update_social_data_format.py]: Drive-by: Are syntax checks passing (`inv syntax`)? This variable appears to be unused.

Comment (sloria) [reply osf/migrations/0124_update_social_data_format.py]: Ah, perhaps uppercased constants don't count as unused variables because it's assumed that they may be imported by other modules.

In any case, I think it can be removed.

File: api/users/serializers.py [hunk 9]
```diff
@@ -34,6 +35,27 @@ def to_representation(self, value):
         return relationship_links
 
 
+class SocialField(ser.DictField):
```
Comment (erinspace) [inline api/users/serializers.py]: Wasn't sure about this approach -- it's not really a deprecated field, as the field is still there, just returns data in a slightly different format depending on the version, which I couldn't find another example of.

File: api/users/serializers.py [hunk 10]
```diff
@@ -34,6 +35,27 @@ def to_representation(self, value):
...
+        self.min_version = min_version
+        self.help_text = 'This field will change data formats after version {}'.format(self.min_version)
+
+    def to_representation(self, value):
+        old_social_string_fields = ['twitter', 'github', 'linkedIn']
+        request = self.context.get('request')
+        show_old_format = request and is_deprecated(request.version, self.min_version)
+        if show_old_format:
+            new_value = {}
+            for key, socialvalue in value.iteritems():
```
Comment (sloria) [inline api/users/serializers.py]: Drive-by: It would be more straight-forward to loop through `old_social_string_field` rather than the entire `social` dictionary.

Comment (sloria) [reply api/users/serializers.py]: Just make sure to make a copy of the social dictionary for the return value (rather than mutating it in-place.

File: website/static/js/profile.js [hunk 11]
```diff
@@ -741,6 +744,17 @@ $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.proto
 
 SocialViewModel.prototype.serialize = function() {
     var serializedData = ko.toJS(this);
+    // Do some extra work to store these as arrays in the DB
+    var arrayInDBKeys = ['twitter', 'linkedIn', 'github'];
+    $.each(serializedData || {}, function(key, value) {
```
Comment (sloria) [inline website/static/js/profile.js]: Same here: more straightforward to loop over the arrayInDBKeys than the whole input dictionary.

Comment (sloria) [reply website/static/js/profile.js]: Also, a nit: we no longer need to use `$.each`, since--as I learned yesterday--we're not supporting browsers older than IE 11, which means `Array.prototype.forEach` is AOK.

File: website/static/js/profile.js [hunk 12]
```diff
@@ -753,11 +767,15 @@ SocialViewModel.prototype.serialize = function() {
 SocialViewModel.prototype.unserialize = function(data) {
     var self = this;
     var websiteValue = [];
+    // Do some extra work to display these as strings for now
+    var arrayInDBKeys = ['twitter', 'linkedIn', 'github'];
```
Comment (sloria) [inline website/static/js/profile.js]: Rather than define this twice, better to define this as a constant.

File: website/static/js/profile.js [hunk 13]
```diff
@@ -739,8 +742,19 @@ var SocialViewModel = function(urls, modes, preventUnsaved) {
 SocialViewModel.prototype = Object.create(BaseViewModel.prototype);
 $.extend(SocialViewModel.prototype, SerializeMixin.prototype, TrackedMixin.prototype);
 
+const arrayInDBKeys = ['twitter', 'linkedIn', 'github'];
```
Comment (sloria) [inline website/static/js/profile.js]: Nit: We're not using Babel to transpile ES2015, so we've just been using `var` (though it's not a huge deal, since `const` is [supported in IE11](https://caniuse.com/#feat=const)). We also use UPPER_SNAKE_CASE for constants.

File: api/users/serializers.py [hunk 14]
```diff
@@ -34,6 +35,27 @@ def to_representation(self, value):
...
+class SocialField(ser.DictField):
+    def __init__(self, min_version, **kwargs):
+        super(SocialField, self).__init__(**kwargs)
+        self.min_version = min_version
+        self.help_text = 'This field will change data formats after version {}'.format(self.min_version)
+
+    def to_representation(self, value):
+        old_social_string_fields = ['twitter', 'github', 'linkedIn']
+        request = self.context.get('request')
+        show_old_format = request and is_deprecated(request.version, self.min_version) and request.method == 'GET'
```
[the remark this answers is not in the document]
Comment (erinspace) [reply api/users/serializers.py]: Ah! So this was because I wasn't exactly sure if versioning was for other operations other than get, and by extension didn't know if we should support patches/puts for data that was in the legacy format. If we indeed are going to support a patch to the social field in the old format, as well as the new format if they're using a current version of the API, that will be a little more work as the jsonschema isn't set up to validate based on version of the API. 

Lemmie know and I can make that happen!

[the remark this answers is not in the document]
Comment (sloria) [reply api/users/serializers.py]: Oh right--non-GET requests were disallowed in older versions. So maybe the `request.method == 'GET'` is just redundant? Can we remove it?

[the remark this answers is not in the document]
Comment (erinspace) [reply api/users/serializers.py]: The reason I added the request.method check was because updates actually secretly **were** allowed in earlier versions -- this feature was actually present already and this PR was just intended to clean it up/make it work the way that's expected. 

This was added primarily to make the [tests in this method](https://github.com/CenterForOpenScience/osf.io/blob/develop/api_tests/users/views/test_user_detail.py#L384) pass while still not specifying the version of the API, cause they were using version 2.0.  I figured since this was a stealth feature before anyhow, it'd be ok to require "proper form" for `PUT`/`POST` requests, while still returning data in the way that's expected for `GET` requests.

**Comment by sloria** [comment]
This is headed to staging3.
